stan/math/rev/functor/ode_bdf_adjoint.hpp

```diff
@@ -0,0 +1,182 @@
+#ifndef STAN_MATH_REV_FUNCTOR_ODE_BDF_ADJOINT_HPP
+#define STAN_MATH_REV_FUNCTOR_ODE_BDF_ADJOINT_HPP
+
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/functor/cvodes_integrator_adjoint.hpp>
+#include <ostream>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Solve the ODE initial value problem y' = f(t, y), y(t0) = y0 at a set of
+ * times, { t1, t2, t3, ... } using the stiff backward differentiation formula
+ * BDF solver from CVODES.
+ *
+ * \p f must define an operator() with the signature as:
+ *   template<typename T_t, typename T_y, typename... T_Args>
+ *   Eigen::Matrix<stan::return_type_t<T_t, T_y, T_Args...>, Eigen::Dynamic, 1>
+ *     operator()(const T_t& t, const Eigen::Matrix<T_y, Eigen::Dynamic, 1>& y,
+ *     std::ostream* msgs, const T_Args&... args);
+ *
+ * t is the time, y is the state, msgs is a stream for error messages, and args
+ * are optional arguments passed to the ODE solve function (which are passed
+ * through to \p f without modification).
+ *
+ * @tparam F Type of ODE right hand side
+ * @tparam T_0 Type of initial time
+ * @tparam T_ts Type of output times
+ * @tparam T_Args Types of pass-through parameters
+ *
+ * @param function_name Calling function name (for printing debugging messages)
+ * @param f Right hand side of the ODE
+ * @param y0 Initial state
+ * @param t0 Initial time
+ * @param ts Times at which to solve the ODE at. All values must be sorted and
+ *   not less than t0.
+ * @param relative_tolerance Relative tolerance passed to CVODES
+ * @param absolute_tolerance Absolute tolerance passed to CVODES
+ * @param max_num_steps Upper limit on the number of integration steps to
+ *   take between each output (error if exceeded)
+ * @param[in, out] msgs the print stream for warning messages
+ * @param args Extra arguments passed unmodified through to ODE right hand side
+ * @return Solution to ODE at times \p ts
+ */
+template <typename F, typename T_y0, typename T_t0, typename T_ts,
+          typename... T_Args, require_eigen_col_vector_t<T_y0>* = nullptr>
+std::vector<Eigen::Matrix<stan::return_type_t<T_y0, T_t0, T_ts, T_Args...>,
+                          Eigen::Dynamic, 1>>
+ode_bdf_adjoint_tol_impl(const char* function_name, const F& f, const T_y0& y0,
+                         const T_t0& t0, const std::vector<T_ts>& ts,
+                         double relative_tolerance, double absolute_tolerance,
+                         long int max_num_steps, std::ostream* msgs,
+                         double absolute_tolerance_B,
+                         double absolute_tolerance_QB,
+                         long int steps_checkpoint, const T_Args&... args) {
+  /*
+  const auto& args_ref_tuple = std::make_tuple(to_ref(args)...);
+  return apply(
+      [&](const auto&... args_refs) {
+        auto integrator
+            = new stan::math::cvodes_integrator_adjoint_vari<CV_BDF, F, T_y0,
+  T_t0, T_ts, T_Args...>( function_name, f, y0, t0, ts, relative_tolerance,
```

Comment:
`T_Args` here will be the types of the args coming in. `args_ref_tuple` however will contain those types after they have been `to_ref` ed.

Reply:
You can use `ref_type_t` to determine the type `to_ref` returns. Similarly `eval_return_type_t` is what `eval` returns. We also have `plain_type_t`, which can also be assigned what `eval` returns, but is also guaranteed not to be a reference (which `eval_return_type_t<T>` is if `T` is already a plain type).

Reply:
Thanks. Is there some doc for this available? Eigen expression for dummies? What you describe sounds all logic, but some overview doc for it would be great (or maybe we have that already and I don't know where). Sorry, but I haven't been part of this process to deploy expressions.

Reply:
Sorry, we don't have such a doc. If you have any questions you can ask me. Or if you have many questions we could talk via hangouts/whatever.

Reply:
Ok... all I need is a resource ... including "living" ones. Great, thanks.